prow/plank/reconciler.go

```diff
@@ -0,0 +1,645 @@
+	if err != nil {
+		return fmt.Errorf("failed to construct predicate: %w", err)
+	}
+
+	if err := mgr.GetFieldIndexer().IndexField(&prowv1.ProwJob{}, prowJobIndexName, prowJobIndexer(cfg().ProwJobNamespace)); err != nil {
+		return fmt.Errorf("failed to add indexer: %w", err)
+	}
+
+	blder := controllerruntime.NewControllerManagedBy(mgr).
+		Named(controllerName).
+		For(&prowv1.ProwJob{}).
+		WithEventFilter(predicate).
+		WithOptions(controller.Options{MaxConcurrentReconciles: numWorkers})
+
+	r := &reconciler{
+		ctx:                context.Background(),
+		pjClient:           mgr.GetClient(),
+		buildClients:       map[string]ctrlruntimeclient.Client{},
+		overwriteReconcile: overwriteReconcile,
+		log:                logrus.NewEntry(logrus.StandardLogger()),
+		config:             cfg,
+		totURL:             totURL,
+		clock:              clock.RealClock{},
+	}
+	for buildCluster, buildClusterMgr := range buildMgrs {
+		blder = blder.Watches(
+			source.NewKindWithCache(&corev1.Pod{}, buildClusterMgr.GetCache()),
+			podEventRequestMapper(cfg().ProwJobNamespace))
+		r.buildClients[buildCluster] = buildClusterMgr.GetClient()
+	}
+
+	if err := blder.Complete(r); err != nil {
+		return fmt.Errorf("failed to build controller: %w", err)
+	}
+
+	if err := mgr.Add(manager.RunnableFunc(r.syncMetrics)); err != nil {
+		return fmt.Errorf("failed to add metrics runnable to manager: %w", err)
+	}
+
+	return nil
+}
+
+type reconciler struct {
+	ctx                context.Context
+	pjClient           ctrlruntimeclient.Client
+	buildClients       map[string]ctrlruntimeclient.Client
+	overwriteReconcile func(reconcile.Request) (reconcile.Result, error)
+	log                *logrus.Entry
+	config             config.Getter
+	totURL             string
+	clock              clock.Clock
+}
+
+func (r *reconciler) syncMetrics(stop <-chan struct{}) error {
+	for {
+		select {
+		case <-stop:
+			return nil
+		case <-time.NewTicker(30 * time.Second).C:
+			pjs := &prowv1.ProwJobList{}
+			if err := r.pjClient.List(r.ctx, pjs, optAllProwJobs()); err != nil {
+				r.log.WithError(err).Error("failed to list prowjobs for metrics")
+				continue
+			}
+			kube.GatherProwJobMetrics(pjs.Items)
+		}
+	}
+}
+
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	if r.overwriteReconcile != nil {
+		return r.overwriteReconcile(request)
+	}
+	return r.defaultReconcile(request)
+}
+
+func (r *reconciler) defaultReconcile(request reconcile.Request) (reconcile.Result, error) {
+	pj := &prowv1.ProwJob{}
+	if err := r.pjClient.Get(r.ctx, request.NamespacedName, pj); err != nil {
+		if !kerrors.IsNotFound(err) {
+			return reconcile.Result{}, fmt.Errorf("failed to get prowjob %s: %v", request.Name, err)
+		}
+
+		// Objects can be deleted from the API while being in our workqueue
+		return reconcile.Result{}, nil
+	}
+
+	// TODO: Terminal errors for unfixable cases like missing build clusters
```

Comment:
As long as those get bubbled up as big fires instead of being swept under the rug by not being retried

Reply:
Yeah they should absolutely be logged but there is no point in retrying them

Reply:
Let me do this in a follow-up, PR is big enough already